Tedlium chain and nnet3 scripts

galv opened the pull request.

danpovey commented:
@vijayaditya, do you have time to check this?

vijayaditya commented:
Sure, will do.

On Sat, Apr 2, 2016 at 5:00 PM, Daniel Povey <notifications@github.com>
wrote:

> @vijayaditya <https://github.com/vijayaditya>, do you have time to check
> this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/644#issuecomment-204804363>
>

vijayaditya commented on egs/tedlium/s5/local/chain/README:
```diff
@@ -0,0 +1,31 @@
+
+Results for each setup are posted at the top of each file.
+
+The rescored test WERs are the same (12.9%), and the second experiment
```
Please remove this comment. There are several reasons for this difference.

galv replied:
if you say so... but the truth is that I ran the training twice for the setup we put in the paper (the second time was to ensure these scripts worked), and one time got 12.9% WER and the other got 12.8% WER. I am quite positive that this is just due to noise in doing stochastic gradient descent.

vijayaditya replied:
There are other reasons for variability e.g. dithering done during feature extraction of test data. Further even the ivector extractors are not the same across the runs. We acknowledge the fact that there is noise in the results generated across different runs. This is true for all experiments in all example recipes. So I don't see any reason for explicitly mentioning this here in particular.

vijayaditya commented on egs/tedlium/s5/local/chain/README:
```diff
@@ -0,0 +1,31 @@
+
+Only after that, you can run:
+
+local/chain/run_tdnn_more_ce.sh
```
This script should be renamed as local/nnet3/run_tdnn.sh

danpovey replied:
Yes, I don't think we need to have two scripts differing so slightly.   If
you want to record the results you could state e.g. in the top of the file,
that when you ran it with standard cross-entropy smoothing value of
cross_entropy_smoothing=0.1 (or whatever it is), you got such-and-such
results.
Dan


On Sat, Apr 2, 2016 at 5:14 PM, Vijayaditya Peddinti <
notifications@github.com> wrote:

> In egs/tedlium/s5/local/chain/README
> <https://github.com/kaldi-asr/kaldi/pull/644#discussion_r58300002>:
>
> > +First run:
> > +
> > +./run.sh
> > +
> > +until the end of stage 7. (local/nnet/run_dnn.sh can be skipped.)
> > +
> > +Then run:
> > +
> > +local/chain/run_tdnn.sh
> > +
> > +to see results for a generic chain model (these were not the results
> > +shown in the paper).
> > +
> > +Only after that, you can run:
> > +
> > +local/chain/run_tdnn_more_ce.sh
>
> This script should be renamed as local/nnet3/run_tdnn.sh
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/644/files/82caa06f43641485acb81b204e02eea21e5bbbeb#r58300002>
>

galv replied:
Sounds good.

vijayaditya commented on egs/tedlium/s5/local/chain/run_tdnn.sh:
```diff
@@ -0,0 +1,216 @@
+# TDNN options
+# this script uses the new tdnn config generator so it needs a final 0 to reflect that the final layer input has no splicing
+# smoothing options
+pool_window=
```
You can remove the pool options, they are not used anymore. Dan recommends removing any variable which is not used more than once.

vijayaditya commented on egs/tedlium/s5/local/chain/run_tdnn.sh:
```diff
@@ -0,0 +1,216 @@
+
+if [ $stage -le 12 ]; then
+  echo "$0: creating neural net configs";
+  if [ ! -z "$relu_dim" ]; then
```
You can remove these options.

vijayaditya commented on egs/tedlium/s5/local/chain/run_tdnn.sh:
```diff
@@ -0,0 +1,216 @@
+    steps/nnet3/decode.sh --acwt 1.0 --post-decode-acwt 10.0 \
+      --nj 8 --cmd "$decode_cmd" $iter_opts \
+      --online-ivector-dir exp/nnet3/ivectors_${decode_set} \
+      --scoring-opts "--decode-mbr false --min-lmwt 8 --max-lmwt 12" \
```
Wasn't there a discussion on using <code>--decode-mbr true</code> in the mail chain ?

galv replied:
Yes. I know how the error cropped up. I will fix this. The paper results used --decode-mbr false, so this is nothing to worry about. I will fix the results posted in the files after the fix.

vijayaditya commented on egs/tedlium/s5/local/chain/run_tdnn.sh:
```diff
@@ -0,0 +1,216 @@
+      --scoring-opts "--decode-mbr false --min-lmwt 8 --max-lmwt 12" \
+      $graph_dir data/${decode_set}_hires $dir/decode_${decode_set}${decode_iter:+_$decode_iter} || exit 1;
+
+    #TODO: Understand: Why doesn't this file take --scoring-opts as an option?
```
Please remove comments to yourself.

vijayaditya commented on egs/tedlium/s5/local/nnet3/run_ivector_common.sh:
```diff
@@ -0,0 +1,113 @@
+# see the chain docs for general direction on what training is doing!
+
+. ./cmd.sh
+set -euo pipefail
```
better not use set -e. It is more informative to print an error message before exit.

vijayaditya commented on egs/tedlium/s5/local/nnet3/run_tdnn.sh:
```diff
@@ -0,0 +1,116 @@
+if [ $stage -le 10 ]; then
+  if [[ $(hostname -f) == *.clsp.jhu.edu ]] && [ ! -d $dir/egs/storage ]; then
+    utils/create_split_dir.pl \
+     /export/b0{3,4,5,6}/$USER/kaldi-data/egs/swbd-$(date +'%m_%d_%H_%M')/s5/$dir/egs/storage $dir/egs/storage
```
use nodes b09,b10,b11,b12

vijayaditya replied:
rename the directory as ...kaldi-data/egs/**tedlium**-...

vijayaditya commented on egs/tedlium/s5/local/nnet3/run_tdnn_discriminative.sh:
```diff
@@ -0,0 +1,211 @@
+
+# Results:
+
+. cmd.sh
```
. ./cmd.sh

vijayaditya commented on egs/tedlium/s5/local/nnet3/run_tdnn_discriminative.sh:
```diff
@@ -0,0 +1,211 @@
+cleanup=false  # run with --cleanup true --stage 6 to clean up (remove large things like denlats,
+               # alignments and degs).
+
+. cmd.sh
```
. ./cmd.sh

vijayaditya commented on egs/tedlium/s5/local/nnet3/run_tdnn_discriminative.sh:
```diff
@@ -0,0 +1,211 @@
+
+dir=${srcdir}_${criterion}
+
+## Egs options
```
remove any variable which is only used once.

danpovey replied:
I'm actually OK with these variables as long as there are just a few of
them and they are carefully chosen, e.g. they point out a couple of the
more commonly-tuned options.    I this case I think it's OK.

On Sat, Apr 2, 2016 at 5:33 PM, Vijayaditya Peddinti <
notifications@github.com> wrote:

> In egs/tedlium/s5/local/nnet3/run_tdnn_discriminative.sh
> <https://github.com/kaldi-asr/kaldi/pull/644#discussion_r58300230>:
>
> > +. ./path.sh
> > +. ./utils/parse_options.sh
> > +
> > +srcdir=exp/nnet3/tdnn_sp
> > +train_data_dir=data/train_sp_hires
> > +online_ivector_dir=exp/nnet3/ivectors_train_sp
> > +degs_dir=                     # If provided, will skip the degs directory creation
> > +lats_dir=                     # If provided, will skip denlats creation
> > +
> > +## Objective options
> > +criterion=smbr
> > +one_silence_class=true
> > +
> > +dir=${srcdir}_${criterion}
> > +
> > +## Egs options
>
> remove any variable which is only used once.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/644/files/82caa06f43641485acb81b204e02eea21e5bbbeb#r58300230>
>

galv replied:
I will be keeping them; it distinguishes what may be reasonable to tune.

vijayaditya replied:
@vimalmanohar Could you please verify this script too..

vimalmanohar replied:
This script looks fine. @galv make sure you are using the same options for decode as the CE baseline. e.g. decode-mbr etc.

vijayaditya replied:
I am assuming @vimalmanohar checked this.

vijayaditya commented on egs/tedlium/s5/local/score_sclite.sh:
```diff
@@ -48,6 +49,15 @@ nj=$(cat $dir/num_jobs)
 
 mkdir -p $dir/scoring/log
 
+if [ -f $dir/../frame_shift ]; then
```
I am not familiar with this part. @vimalmanohar could you please review this script.

vijayaditya commented on egs/tedlium/s5/local/chain/run_tdnn.sh:
```diff
@@ -0,0 +1,216 @@
+  fi
+
+  # create the config files for nnet initialization
+  pool_opts=
```
We are not using pooling any more. You can remove these options.

danpovey replied:
vijay, perhaps you could create an issue to remove the pooling from other
scripts, if you're not planning to make more use of that?

On Sat, Apr 2, 2016 at 5:41 PM, Vijayaditya Peddinti <
notifications@github.com> wrote:

> In egs/tedlium/s5/local/chain/run_tdnn.sh
> <https://github.com/kaldi-asr/kaldi/pull/644#discussion_r58300350>:
>
> > +  # Build a tree using our new topology.
> > +  steps/nnet3/chain/build_tree.sh --frame-subsampling-factor 3 \
> > +      --leftmost-questions-truncate $leftmost_questions_truncate \
> > +      --cmd "$train_cmd" 4200 data/$train_set $lang $ali_dir $treedir
> > +fi
> > +
> > +if [ $stage -le 12 ]; then
> > +  echo "$0: creating neural net configs";
> > +  if [ ! -z "$relu_dim" ]; then
> > +    dim_opts="--relu-dim $relu_dim"
> > +  else
> > +    dim_opts="--pnorm-input-dim $pnorm_input_dim --pnorm-output-dim  $pnorm_output_dim"
> > +  fi
> > +
> > +  # create the config files for nnet initialization
> > +  pool_opts=
>
> We are not using pooling any more. You can remove these options.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/644/files/82caa06f43641485acb81b204e02eea21e5bbbeb#r58300350>
>

vijayaditya replied:
@danpovey raised an issue. Will address it this week.

vijayaditya commented on egs/tedlium/s5/local/chain/run_tdnn.sh:
```diff
@@ -0,0 +1,216 @@
+pool_lpfilter_width=
+self_repair_scale=0.00001
+# training options
+num_epochs=4
```
You are using num_epochs=2 for xent training and num_epochs=4 for chain training. Training for n epochs with chain objective is equivalent to training for 3n epochs, as we perform frame-shifts. So I would recommend reducing the num_epochs. Also given that you were training for so many epochs I would recommend plotting the train/validation objfn values and checking if there is anything unexpected.

danpovey replied:
It might be worth trying to train the TDNN system for more epochs and see
if it improves, for due diligence-- it might improve.
Dan


On Sat, Apr 2, 2016 at 5:49 PM, Vijayaditya Peddinti <
notifications@github.com> wrote:

> In egs/tedlium/s5/local/chain/run_tdnn.sh
> <https://github.com/kaldi-asr/kaldi/pull/644#discussion_r58300469>:
>
> > +stage=0 # After running the entire script once, you can set stage=12 to tune the neural net only.
> > +train_stage=-10
> > +get_egs_stage=-10
> > +speed_perturb=true
> > +dir=exp/chain/tdnn  # Note: _sp will get added to this if $speed_perturb == true.
> > +decode_iter=
> > +
> > +# TDNN options
> > +# this script uses the new tdnn config generator so it needs a final 0 to reflect that the final layer input has no splicing
> > +# smoothing options
> > +pool_window=
> > +pool_type='none'
> > +pool_lpfilter_width=
> > +self_repair_scale=0.00001
> > +# training options
> > +num_epochs=4
>
> You are using num_epochs=2 for xent training and num_epochs=4 for chain
> training. Training for n epochs with chain objective is equivalent to
> training for 3n epochs, as we perform frame-shifts. So I would recommend
> reducing the num_epochs. Also given that you were training for so many
> epochs I would recommend plotting the train/validation objfn values and
> checking if there is anything unexpected.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/644/files/82caa06f43641485acb81b204e02eea21e5bbbeb#r58300469>
>

vimalmanohar commented on egs/tedlium/s5/local/score_sclite.sh:
```diff
@@ -68,8 +78,8 @@ fi
-    cat $x | grep -v -E '"\[BREATH|NOISE|COUGH|SMACK|UM|UH\]"' | \
-      grep -v -E '"!SIL|\<UNK\>"' > ${x}.filt || exit 1;
+    cat $x | grep -v -w -E '\[BREATH|NOISE|COUGH|SMACK|UM|UH\]' | \
+      grep -v -w -E '!SIL|<unk>' > ${x}.filt || exit 1;
```
It might make sense to add -i option to ignore case.

galv replied:
I will do this; it is more robust in case someone copies this code without thinking enough.

vimalmanohar commented on egs/tedlium/s5/local/score_sclite.sh:
```diff
@@ -13,6 +13,7 @@ beam=7  # speed-up, but may affect MBR confidences.
 word_ins_penalty=0.0,0.5,1.0
 min_lmwt=10
 max_lmwt=20
+iter=final
```
You need to replace every occurence of final.mdl with $iter.mdl

drTonyRobinson commented:
On 02/04/16 21:55, Daniel Galvez wrote:
> tedlium: Make scoring scripts correctly remove <unk> from ctm files.

Should <unk> even get into ctm files?

IIRC the pronunciation dictionary we provided had a bug which meant that 
<unk> had a pronunciation (probably "ah n k") and so was decoded by 
default, this got fixed and Dan commented that the WER improved 
considerably.


Tony

-- 
Speechmatics is a trading name of Cantab Research Limited
We are hiring: www.speechmatics.com/careers 
<https://www.speechmatics.com/careers>
Dr A J Robinson, Founder, Cantab Research Ltd
Phone direct: 01223 794096, office: 01223 794497
Company reg no GB 05697423, VAT reg no 925606030
51 Canterbury Street, Cambridge, CB4 3QG, UK

danpovey commented:
It *can* get in there because by default we allow it to be decoded
(pronunciation is the spoken-noise phone's pronunciation, it gets fed by
OOV words and other bad things).  Usually it's rarely decoded, but in the
chain models it can be decoded more often than is desirable, so we recently
added (in chain branch) and option --remove-oov to mkgraph.sh that will
make sure it doesn't appear in the graph.
Dan


On Sat, Apr 2, 2016 at 10:36 PM, drTonyRobinson <notifications@github.com>
wrote:

> On 02/04/16 21:55, Daniel Galvez wrote:
> > tedlium: Make scoring scripts correctly remove <unk> from ctm files.
>
> Should <unk> even get into ctm files?
>
> IIRC the pronunciation dictionary we provided had a bug which meant that
> <unk> had a pronunciation (probably "ah n k") and so was decoded by
> default, this got fixed and Dan commented that the WER improved
> considerably.
>
>
> Tony
>
> --
> Speechmatics is a trading name of Cantab Research Limited
> We are hiring: www.speechmatics.com/careers
> <https://www.speechmatics.com/careers>
> Dr A J Robinson, Founder, Cantab Research Ltd
> Phone direct: 01223 794096, office: 01223 794497
> Company reg no GB 05697423, VAT reg no 925606030
> 51 Canterbury Street, Cambridge, CB4 3QG, UK
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/644#issuecomment-204857603>
>

drTonyRobinson commented:
Thanks.  I just wanted to make sure that a bug we put in wasn't reemerging.


Tony

On 03/04/16 03:46, Daniel Povey wrote:
> It *can* get in there because by default we allow it to be decoded
> (pronunciation is the spoken-noise phone's pronunciation, it gets fed by
> OOV words and other bad things). Usually it's rarely decoded, but in the
> chain models it can be decoded more often than is desirable, so we 
> recently
> added (in chain branch) and option --remove-oov to mkgraph.sh that will
> make sure it doesn't appear in the graph.
> Dan
>
>
> On Sat, Apr 2, 2016 at 10:36 PM, drTonyRobinson <notifications@github.com>
> wrote:
>
> > On 02/04/16 21:55, Daniel Galvez wrote:
> > > tedlium: Make scoring scripts correctly remove <unk> from ctm files.
> >
> > Should <unk> even get into ctm files?
> >
> > IIRC the pronunciation dictionary we provided had a bug which meant that
> > <unk> had a pronunciation (probably "ah n k") and so was decoded by
> > default, this got fixed and Dan commented that the WER improved
> > considerably.
> >
> >
> > Tony
> >
> > --
> > Speechmatics is a trading name of Cantab Research Limited
> > We are hiring: www.speechmatics.com/careers
> > <https://www.speechmatics.com/careers>
> > Dr A J Robinson, Founder, Cantab Research Ltd
> > Phone direct: 01223 794096, office: 01223 794497
> > Company reg no GB 05697423, VAT reg no 925606030
> > 51 Canterbury Street, Cambridge, CB4 3QG, UK
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly or view it on GitHub
> > <https://github.com/kaldi-asr/kaldi/pull/644#issuecomment-204857603>
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub 
> <https://github.com/kaldi-asr/kaldi/pull/644#issuecomment-204858422>
>


-- 
Speechmatics is a trading name of Cantab Research Limited
We are hiring: www.speechmatics.com/careers 
<https://www.speechmatics.com/careers>
Dr A J Robinson, Founder, Cantab Research Ltd
Phone direct: 01223 794096, office: 01223 794497
Company reg no GB 05697423, VAT reg no 925606030
51 Canterbury Street, Cambridge, CB4 3QG, UK

galv commented:
Thanks for the feedback, everyone. I will have to look at this seriously tomorrow; I have been traveling all day. I have one question: Is modifying these current commits okay? (I believe github's UI may remove comments on commits that get retroactively modified.) I do not want to put "broken" commits into a main branch. (e.g., Vimal's comment on not using the variable $iter. I thought this variable was already being used, as my decoding results are different for each eoch of discriminative training. I will investigate.)

danpovey commented:
I think it's OK to do non-fast-forward updates for pull requests, github
does its best to work out the correct location for comments.

On Sun, Apr 3, 2016 at 11:23 PM, Daniel Galvez <notifications@github.com>
wrote:

> Thanks for the feedback, everyone. I will have to look at this seriously
> tomorrow; I have been traveling all day. I have one question: Is modifying
> these current commits okay? (I believe github's UI may remove comments on
> commits that get retroactively modified.) I do not want to put "broken"
> commits into a main branch. (e.g., Vimal's comment on not using the
> variable $iter. I thought this variable was already being used, as my
> decoding results are different for each eoch of discriminative training. I
> will investigate.)
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/644#issuecomment-205117453>
>

galv commented:
It is okay to force push okay; I had simply gotten confused by an also confused friend.

Fixes are made; I still need to wait on decoding to finish for models using "--decode-mbr true" (to be clear, the model in the paper uses this option. This just snuck in while I was trying to clean up; I just need to redo decoding for the model whose results are reported here, which is distinct from the model in the paper in its results (same hyperparameters).)

galv commented:
I decoded without the accidental "--decode-mbr false". The model with extra cross entropy regularization now matches results int he paper: WER 12.8% rescored on test set (full results at top of local/chain/run_tdnn.sh). :+1:

I have not looked into Vijay's comment on changing the number of epochs or looked at how training or validation curves do, yet. Is using "--remove-oov" something we want to do here? I cannot remember for what setup we wanted that option. This discussion can take place via email.

Otherwise, these commits should be good to go for reproducing the paper's results.

vijayaditya commented:
The paper results are not yet final. We will keep updating these results as
we improve them. So just choose the best setups.

On Tue, Apr 5, 2016 at 12:02 AM, Daniel Galvez <notifications@github.com>
wrote:

> I decoded without the accidental "--decode-mbr false". The model with
> extra cross entropy regularization now matches results int he paper: WER
> 12.8% rescored on test set (full results at top of
> local/chain/run_tdnn.sh). [image: :+1:]
>
> I have not looked into Vijay's comment on changing the number of epochs or
> looked at how training or validation curves do, yet. Is using
> "--remove-oov" something we want to do here? I cannot remember for what
> setup we wanted that option. This discussion can take place via email.
>
> Otherwise, these commits should be good to go for reproducing the paper's
> results.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/644#issuecomment-205633105>
>

danpovey commented:
--remove-oov is something you can do in graph creation [option to
mkgraph.sh] if you are excessively decoding `<UNK>` (which the chain models
tend to do).  But this only works if `<UNK>` (or `<unk>`, depending on the
corpus) is present in data/lang/oov.txt.  It should be, but in librispeech,
it wasn't.
If you're not getting tons of sentences with unk in them, it wouldn't make
a difference.
Dan


On Tue, Apr 5, 2016 at 12:10 AM, Vijayaditya Peddinti <
notifications@github.com> wrote:

> The paper results are not yet final. We will keep updating these results as
> we improve them. So just choose the best setups.
>
> On Tue, Apr 5, 2016 at 12:02 AM, Daniel Galvez <notifications@github.com>
> wrote:
>
> > I decoded without the accidental "--decode-mbr false". The model with
> > extra cross entropy regularization now matches results int he paper: WER
> > 12.8% rescored on test set (full results at top of
> > local/chain/run_tdnn.sh). [image: :+1:]
> >
> > I have not looked into Vijay's comment on changing the number of epochs
> or
> > looked at how training or validation curves do, yet. Is using
> > "--remove-oov" something we want to do here? I cannot remember for what
> > setup we wanted that option. This discussion can take place via email.
> >
> > Otherwise, these commits should be good to go for reproducing the paper's
> > results.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly or view it on GitHub
> > <https://github.com/kaldi-asr/kaldi/pull/644#issuecomment-205633105>
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/644#issuecomment-205634483>
>

galv commented:
@vijayaditya You're right; I had not thought about the camera ready paper. What I meant was that I think that this PR can be merged now, as it can reproduce our current results. At the very least, it would be good to get the earlier commits in soon, as they are not limited exclusively to the chain models.

danpovey commented:
VIjay, merge this if you think it's ready- I am too busy today.

On Tue, Apr 5, 2016 at 12:48 AM, Daniel Galvez <notifications@github.com>
wrote:

> @vijayaditya <https://github.com/vijayaditya> You're right; I had not
> thought about the camera ready paper. What I meant was that I think that
> this PR can be merged now, as it can reproduce our current results. At the
> very least, it would be good to get the earlier commits in soon, as they
> are not limited exclusively to the chain models.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/644#issuecomment-205643111>
>

vijayaditya commented:
OK, I will review it later today.

--Vijay

On Tue, Apr 5, 2016 at 1:43 PM, Daniel Povey <notifications@github.com>
wrote:

> VIjay, merge this if you think it's ready- I am too busy today.
>
> On Tue, Apr 5, 2016 at 12:48 AM, Daniel Galvez <notifications@github.com>
> wrote:
>
> > @vijayaditya <https://github.com/vijayaditya> You're right; I had not
> > thought about the camera ready paper. What I meant was that I think that
> > this PR can be merged now, as it can reproduce our current results. At
> the
> > very least, it would be good to get the earlier commits in soon, as they
> > are not limited exclusively to the chain models.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly or view it on GitHub
> > <https://github.com/kaldi-asr/kaldi/pull/644#issuecomment-205643111>
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/644#issuecomment-205914993>
>

vijayaditya commented on egs/tedlium/s5/local/nnet3/run_tdnn.sh:
```diff
@@ -0,0 +1,113 @@
+    --egs.dir "$common_egs_dir" \
+    --cleanup.remove-egs $remove_egs \
+    --cleanup.preserve-model-interval 100 \
+    --use-gpu true \
```
I think there is no point explicitly specifying use-gpu true as we never train without GPUs.

naxingyu replied:
Vij, I'll take your comments in this PR into mine too :)

Xingyu

在 2016/4/6 12:03, Vijayaditya Peddinti 写道:
>
> In egs/tedlium/s5/local/nnet3/run_tdnn.sh 
> <https://github.com/kaldi-asr/kaldi/pull/644#discussion_r58648507>:
>
> > +     /export/b{09,10,11,12}/$USER/kaldi-data/egs/tedlium-$(date +'%m_%d_%H_%M')/s5/$dir/egs/storage $dir/egs/storage
> > +  fi
> > +
> > +  steps/nnet3/train_dnn.py --stage=$train_stage \
> > +    --cmd="$decode_cmd" \
> > +    --feat.online-ivector-dir exp/nnet3/ivectors_${train_set} \
> > +    --feat.cmvn-opts="--norm-means=false --norm-vars=false" \
> > +    --trainer.num-epochs 2 \
> > +    --trainer.optimization.num-jobs-initial 3 \
> > +    --trainer.optimization.num-jobs-final 8 \
> > +    --trainer.optimization.initial-effective-lrate 0.0015 \
> > +    --trainer.optimization.final-effective-lrate 0.00015 \
> > +    --egs.dir "$common_egs_dir" \
> > +    --cleanup.remove-egs $remove_egs \
> > +    --cleanup.preserve-model-interval 100 \
> > +    --use-gpu true \
>
> I think there is no point explicitly specifying use-gpu true as we 
> never train without GPUs.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/kaldi-asr/kaldi/pull/644/files/036396efc35ad2cdb5b2039fdfce6b8e9f3d5786#r58648507>
>

vijayaditya commented on egs/tedlium/s5/local/nnet3/run_tdnn.sh:
```diff
@@ -0,0 +1,113 @@
+# If you want to run without GPU you'd have to call train_tdnn.sh with --gpu false,
+# --num-threads 16 and --minibatch-size 128.
+
+stage=9
```
better use stage=1 for this script.

galv replied:
That could have been bad :/

vijayaditya commented on egs/tedlium/s5/local/nnet3/run_tdnn.sh:
```diff
@@ -0,0 +1,113 @@
+stage=9
+affix=
+train_stage=-10
+speed_perturb=true
```
We are using speed-perturbation in all the numbers reported in recipe. There is no point keeping this as a configurable option anymore.

galv replied:
That sounds reasonable. What do you think of the _sp suffix? I am not sure whether or not we should keep it or not.

vijayaditya replied:
since we are rewriting all the recipes now would be the time to drop it and still maintain consistency in naming across all recipes. So @freewym @galv @naxingyu you can drop the _sp suffix from the experiment directory name.

naxingyu replied:
ok

在 2016/4/7 13:07, Vijayaditya Peddinti 写道:
>
> In egs/tedlium/s5/local/nnet3/run_tdnn.sh 
> <https://github.com/kaldi-asr/kaldi/pull/644#discussion_r58819713>:
>
> > +# call multi-splice.
> > +
> > +# Results:
> > +# %WER 15.3 | 507 17792 | 87.4 9.0 3.6 2.7 15.3 90.1 | -0.081 | exp/nnet3/tdnn_sp/decode_dev/score_10_0.5/ctm.filt.filt.sys
> > +# %WER 13.9 | 507 17792 | 88.4 8.0 3.6 2.3 13.9 85.8 | -0.164 | exp/nnet3/tdnn_sp/decode_dev_rescore/score_10_0.5/ctm.filt.filt.sys
> > +# %WER 13.8 | 1155 27512 | 88.5 8.7 2.7 2.3 13.8 84.2 | -0.076 | exp/nnet3/tdnn_sp/decode_test/score_10_0.0/ctm.filt.filt.sys
> > +# %WER 12.5 | 1155 27512 | 89.6 7.7 2.6 2.1 12.5 81.5 | -0.133 | exp/nnet3/tdnn_sp/decode_test_rescore/score_10_0.0/ctm.filt.filt.sys
> > +
> > +# At this script level we don't support not running on GPU, as it would be painfully slow.
> > +# If you want to run without GPU you'd have to call train_tdnn.sh with --gpu false,
> > +# --num-threads 16 and --minibatch-size 128.
> > +
> > +stage=9
> > +affix=
> > +train_stage=-10
> > +speed_perturb=true
>
> since we are rewriting all the recipes now would be the time to drop 
> it and still maintain consistency in naming across all recipes. So 
> @freewym <https://github.com/freewym> @galv <https://github.com/galv> 
> @naxingyu <https://github.com/naxingyu> you can drop the _sp suffix 
> from the experiment directory name.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub 
> <https://github.com/kaldi-asr/kaldi/pull/644/files/036396efc35ad2cdb5b2039fdfce6b8e9f3d5786#r58819713>
>

naxingyu replied:
Only the nnet experiment dir name, or all related experiment dirs 
including ali, lats and chain tree?

在 2016/4/7 13:07, Vijayaditya Peddinti 写道:
>
> In egs/tedlium/s5/local/nnet3/run_tdnn.sh 
> <https://github.com/kaldi-asr/kaldi/pull/644#discussion_r58819713>:
>
> > +# call multi-splice.
> > +
> > +# Results:
> > +# %WER 15.3 | 507 17792 | 87.4 9.0 3.6 2.7 15.3 90.1 | -0.081 | exp/nnet3/tdnn_sp/decode_dev/score_10_0.5/ctm.filt.filt.sys
> > +# %WER 13.9 | 507 17792 | 88.4 8.0 3.6 2.3 13.9 85.8 | -0.164 | exp/nnet3/tdnn_sp/decode_dev_rescore/score_10_0.5/ctm.filt.filt.sys
> > +# %WER 13.8 | 1155 27512 | 88.5 8.7 2.7 2.3 13.8 84.2 | -0.076 | exp/nnet3/tdnn_sp/decode_test/score_10_0.0/ctm.filt.filt.sys
> > +# %WER 12.5 | 1155 27512 | 89.6 7.7 2.6 2.1 12.5 81.5 | -0.133 | exp/nnet3/tdnn_sp/decode_test_rescore/score_10_0.0/ctm.filt.filt.sys
> > +
> > +# At this script level we don't support not running on GPU, as it would be painfully slow.
> > +# If you want to run without GPU you'd have to call train_tdnn.sh with --gpu false,
> > +# --num-threads 16 and --minibatch-size 128.
> > +
> > +stage=9
> > +affix=
> > +train_stage=-10
> > +speed_perturb=true
>
> since we are rewriting all the recipes now would be the time to drop 
> it and still maintain consistency in naming across all recipes. So 
> @freewym <https://github.com/freewym> @galv <https://github.com/galv> 
> @naxingyu <https://github.com/naxingyu> you can drop the _sp suffix 
> from the experiment directory name.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub 
> <https://github.com/kaldi-asr/kaldi/pull/644/files/036396efc35ad2cdb5b2039fdfce6b8e9f3d5786#r58819713>
>

danpovey replied:
The alignment directories should probably keep the _sp suffix because we
can potentially create them with the same model as other alignment
directories with the non-speed-perturbed data.
Personally I would be OK to keep the _sp suffix everywhere but I don't feel
strongly about it.

Dan


On Mon, Apr 11, 2016 at 7:03 PM, Xingyu Na <notifications@github.com> wrote:

> In egs/tedlium/s5/local/nnet3/run_tdnn.sh
> <https://github.com/kaldi-asr/kaldi/pull/644#discussion_r59311717>:
>
> > +# call multi-splice.
> > +
> > +# Results:
> > +# %WER 15.3 | 507 17792 | 87.4 9.0 3.6 2.7 15.3 90.1 | -0.081 | exp/nnet3/tdnn_sp/decode_dev/score_10_0.5/ctm.filt.filt.sys
> > +# %WER 13.9 | 507 17792 | 88.4 8.0 3.6 2.3 13.9 85.8 | -0.164 | exp/nnet3/tdnn_sp/decode_dev_rescore/score_10_0.5/ctm.filt.filt.sys
> > +# %WER 13.8 | 1155 27512 | 88.5 8.7 2.7 2.3 13.8 84.2 | -0.076 | exp/nnet3/tdnn_sp/decode_test/score_10_0.0/ctm.filt.filt.sys
> > +# %WER 12.5 | 1155 27512 | 89.6 7.7 2.6 2.1 12.5 81.5 | -0.133 | exp/nnet3/tdnn_sp/decode_test_rescore/score_10_0.0/ctm.filt.filt.sys
> > +
> > +# At this script level we don't support not running on GPU, as it would be painfully slow.
> > +# If you want to run without GPU you'd have to call train_tdnn.sh with --gpu false,
> > +# --num-threads 16 and --minibatch-size 128.
> > +
> > +stage=9
> > +affix=
> > +train_stage=-10
> > +speed_perturb=true
>
> Only the nnet experiment dir name, or all related experiment dirs
> including ali, lats and chain tree? 在 2016/4/7 13:07, Vijayaditya Peddinti
> 写道:
> … <#m_5858000871164940598_>
> In egs/tedlium/s5/local/nnet3/run_tdnn.sh <#644 (comment)
> <https://github.com/kaldi-asr/kaldi/pull/644#discussion_r58819713>>: > +#
> call multi-splice. > + > +# Results: > +# %WER 15.3 | 507 17792 | 87.4 9.0
> 3.6 2.7 15.3 90.1 | -0.081 |
> exp/nnet3/tdnn_sp/decode_dev/score_10_0.5/ctm.filt.filt.sys > +# %WER 13.9
> | 507 17792 | 88.4 8.0 3.6 2.3 13.9 85.8 | -0.164 |
> exp/nnet3/tdnn_sp/decode_dev_rescore/score_10_0.5/ctm.filt.filt.sys > +#
> %WER 13.8 | 1155 27512 | 88.5 8.7 2.7 2.3 13.8 84.2 | -0.076 |
> exp/nnet3/tdnn_sp/decode_test/score_10_0.0/ctm.filt.filt.sys > +# %WER 12.5
> | 1155 27512 | 89.6 7.7 2.6 2.1 12.5 81.5 | -0.133 |
> exp/nnet3/tdnn_sp/decode_test_rescore/score_10_0.0/ctm.filt.filt.sys > + >
> +# At this script level we don't support not running on GPU, as it would be
> painfully slow. > +# If you want to run without GPU you'd have to call
> train_tdnn.sh with --gpu false, > +# --num-threads 16 and --minibatch-size
> 128. > + > +stage=9 > +affix= > +train_stage=-10 > +speed_perturb=true
> since we are rewriting all the recipes now would be the time to drop it and
> still maintain consistency in naming across all recipes. So @freewym
> <https://github.com/freewym> <https://github.com/freewym> @galv
> <https://github.com/galv> <https://github.com/galv> @naxingyu
> <https://github.com/naxingyu> <https://github.com/naxingyu> you can drop
> the _sp suffix from the experiment directory name. — You are receiving this
> because you were mentioned. Reply to this email directly or view it on
> GitHub <
> https://github.com/kaldi-asr/kaldi/pull/644/files/036396efc35ad2cdb5b2039fdfce6b8e9f3d5786#r58819713>
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/644/files/036396efc35ad2cdb5b2039fdfce6b8e9f3d5786#r59311717>
>

vijayaditya commented on egs/tedlium/s5/local/nnet3/run_tdnn.sh:
```diff
@@ -0,0 +1,113 @@
+speed_perturb=true
+common_egs_dir=
+reporting_email=
+remove_egs=false
```
remove_egs=true

galv commented:
@vijayaditya You told Yiming to make speed perturbation required in #645. Should I do the same here?

vijayaditya commented on egs/tedlium/s5/local/nnet3/run_tdnn.sh:
```diff
@@ -0,0 +1,113 @@
+    --cmd="$decode_cmd" \
+    --feat.online-ivector-dir exp/nnet3/ivectors_${train_set} \
+    --feat.cmvn-opts="--norm-means=false --norm-vars=false" \
+    --trainer.num-epochs 2 \
```
did you run an experiment with larger number of epochs ? Did it help the baseline system ?

galv replied:
Running the baseline with 6 epochs of training now. chain system, while its validation error never went down, did not see its validation error go up in the last 1.5 epochs of the 4. I am rerunning with --num-epochs 3 for the chain model. I am running with model preserve interval of 20 so I get a finer sense of how decoding does at various points.

vijayaditya commented on egs/tedlium/s5/local/nnet3/run_tdnn.sh:
```diff
@@ -0,0 +1,113 @@
+    --trainer.optimization.final-effective-lrate 0.00015 \
+    --egs.dir "$common_egs_dir" \
+    --cleanup.remove-egs $remove_egs \
+    --cleanup.preserve-model-interval 100 \
```
better make this interval 20 or 30. I don't think this script runs for a large number of iterations.

vijayaditya commented on egs/tedlium/s5/local/nnet3/run_tdnn.sh:
```diff
@@ -0,0 +1,113 @@
+    --ivector-dir exp/nnet3/ivectors_${train_set} \
+    --ali-dir $ali_dir \
+    --relu-dim 500 \
+    --splice-indexes "-1,0,1 -1,0,1,2 -3,0,3 -3,0,3 -3,0,3 -6,-3,0" \
```
are these splice indices correct ? Did you find them better than the baseline TDNN in local/online/run_nnet2_ms.sh ?

danpovey replied:
vijay, just a side note: I think when I last tuned these splice indexes, we were still using the jesus-layer.  The jesus-layer was more amenable to have >2 splice indexes per layer because the parameters don't increase linearly with the number of splice indexes, like they do with TDNNs.  So it's possible that in the TDNN systems,  skipping the middle indexes and maybe having the separation go from 3 to 6 with higher layers might be helpful (along with increasing the relu-dim slightly).  Maybe you could have someone experiment with this.

vijayaditya replied:
@danpovey Sorry I just saw this comment. I ran some experiments where I was always using 0th time index and I was splicing only a part of the hidden vector from the non-zero splice indexes. This gave slightly better results in AMI. We could try and repeat these experiments for the chain and nnet3 models. I will add it to the to-do list.

vijayaditya commented on egs/tedlium/s5/local/chain/run_tdnn.sh:
```diff
@@ -0,0 +1,211 @@
+mkdir -p $dir
+# if we are using the speed-perturbed data we need to generate
+# alignments for it.
+local/nnet3/run_ivector_common.sh --stage $stage \
```
You need not generate alignments for chain training. You can reuse the alignments from the HMM-GMM recipe  for tree building. Modify the directories in the tree building stage appropriately.

vijayaditya commented on egs/tedlium/s5/local/chain/run_tdnn.sh:
```diff
@@ -0,0 +1,211 @@
+gmm_dir=exp/$gmm
+ali_dir=exp/${gmm}_ali$suffix
+lats_dir=${ali_dir/ali/lats}
+treedir=exp/chain/tri3_tree$suffix
```
if you want to keep $gmm. Use it even while setting treedir.

vijayaditya commented on egs/tedlium/s5/local/chain/run_tdnn.sh:
```diff
@@ -0,0 +1,211 @@
+dir=exp/chain/tdnn  # Note: _sp will get added to this if $speed_perturb == true.
+decode_iter=
+
+# TDNN options
```
I would remove all these variables.

vijayaditya commented on egs/tedlium/s5/local/chain/run_tdnn.sh:
```diff
@@ -0,0 +1,211 @@
+   --cleanup.remove-egs $remove_egs \
+   --feat-dir data/${train_set}_hires \
+   --tree-dir $treedir \
+   --lat-dir exp/tri3_lats$suffix \
```
once again .. if you are using a variable $gmm then use it in place of every tri3 occurrence in this script.

vijayaditya commented:
Yes.

On Wed, Apr 6, 2016 at 12:09 AM, Daniel Galvez <notifications@github.com>
wrote:

> @vijayaditya <https://github.com/vijayaditya> You told Yiming to make
> speed perturbation required in #645
> <https://github.com/kaldi-asr/kaldi/pull/645>. Should I do the same here?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/644#issuecomment-206110403>
>

vijayaditya commented on egs/tedlium/s5/local/chain/run_tdnn.sh:
```diff
@@ -0,0 +1,211 @@
+
+if [ $stage -le 12 ]; then
+  echo "$0: creating neural net configs";
+  dim_opts="--relu-dim $relu_dim"
```
there is no point keeping this variable anymore.

vijayaditya commented on egs/tedlium/s5/local/chain/run_tdnn.sh:
```diff
@@ -0,0 +1,211 @@
+    steps/nnet3/decode.sh --acwt 1.0 --post-decode-acwt 10.0 \
+      --nj 8 --cmd "$decode_cmd" $iter_opts \
+      --online-ivector-dir exp/nnet3/ivectors_${decode_set} \
+      --scoring-opts "--min-lmwt 8 --max-lmwt 12" \
```
did you verify that the best lmwt is between 8-12 ? If not you should increase it to 5-15.

vijayaditya commented:
@galv When you are done with the changes, please send a message.

galv commented:
While updating the code based on Vijay's suggestions, I noticed that wav.scp has very little granularity: There is a wav file for every TED talk. This means that we volume-perturb on the granularity of an entire speech, which seems bad. I am running a chain experiment without volume perturbation now. Is there an easy utility script to split a wav file into multiple ones in Kaldi? If not, I'll just put something together myself using sox. I want to investigate perturbing shorter utterances.

I also spotted a mistake: I volume-perturbed the dev and test data. I have fixed this.

Training for 6 epochs a baseline model gives increased WER on the test set, but reduced WER on the development set:

#Training for 2 epochs
# %WER 15.3 | 507 17792 | 87.4 9.0 3.6 2.7 15.3 90.1 | -0.081 | exp/nnet3/tdnn_sp/decode_dev/score_10_0.5/ctm.filt.filt.sys
# %WER 13.9 | 507 17792 | 88.4 8.0 3.6 2.3 13.9 85.8 | -0.164 | exp/nnet3/tdnn_sp/decode_dev_rescore/score_10_0.5/ctm.filt.filt.sys
# %WER 13.8 | 1155 27512 | 88.5 8.7 2.7 2.3 13.8 84.2 | -0.076 | exp/nnet3/tdnn_sp/decode_test/score_10_0.0/ctm.filt.filt.sys                                                                                      
# %WER 12.5 | 1155 27512 | 89.6 7.7 2.6 2.1 12.5 81.5 | -0.133 | exp/nnet3/tdnn_sp/decode_test_rescore/score_10_0.0/ctm.filt.filt.sys                                                                              

# Training for 6 epochs.                                             
#%WER 15.2 | 507 17792 | 87.2 8.9 3.9 2.5 15.2 89.3 | -0.113 | exp/nnet3/tdnn_6epochs/decode_dev/score_10_1.0/ctm.filt.filt.sys                                                                                    
#%WER 13.8 | 507 17792 | 89.0 8.1 2.9 2.9 13.8 87.0 | -0.144 | exp/nnet3/tdnn_6epochs/decode_dev_rescore/score_10_0.0/ctm.filt.filt.sys                                                                            
#%WER 14.0 | 1155 27512 | 88.4 8.8 2.8 2.4 14.0 84.2 | -0.161 | exp/nnet3/tdnn_6epochs/decode_test/score_10_0.0/ctm.filt.filt.sys                                                                                  
#%WER 12.8 | 1155 27512 | 89.3 8.0 2.7 2.2 12.8 82.1 | -0.171 | exp/nnet3/tdnn_6epochs/decode_test_rescore/score_10_0.0/ctm.filt.filt.sys                                                                          

This might reflect the difference in transcription quality between {train,dev} and test sets; it could be overfitting to the poorer transcription. I will need to rerun the baseline without the messed up volume perturbation to be sure. (Waiting for the ivectors to be generated right now.)

vijayaditya commented:
We usually have wave files for entire conversations I.e., one file for 10
minutes. How many speakers are there in Ted and what is the average
duration of these talks ?

Vijay
On Apr 9, 2016 22:44, "Daniel Galvez" <notifications@github.com> wrote:

> While updating the code based on Vijay's suggestions, I noticed that
> wav.scp has very little granularity: There is a wav file for every TED
> talk. This means that we volume-perturb on the granularity of an entire
> speech, which seems bad. I am running a chain experiment without volume
> perturbation now. Is there an easy utility script to split a wav file into
> multiple ones in Kaldi? If not, I'll just put something together myself
> using sox. I want to investigate perturbing shorter utterances.
>
> I also spotted a mistake: I volume-perturbed the dev and test data. I have
> fixed this.
>
> Training for 6 epochs a baseline model gives increased WER on the test
> set, but reduced WER on the development set:
>
> #Training for 2 epochs
> %WER 15.3 | 507 17792 | 87.4 9.0 3.6 2.7 15.3 90.1 | -0.081 |
> exp/nnet3/tdnn_sp/decode_dev/score_10_0.5/ctm.filt.filt.sys %WER 13.9 |
> 507 17792 | 88.4 8.0 3.6 2.3 13.9 85.8 | -0.164 |
> exp/nnet3/tdnn_sp/decode_dev_rescore/score_10_0.5/ctm.filt.filt.sys %WER
> 13.8 | 1155 27512 | 88.5 8.7 2.7 2.3 13.8 84.2 | -0.076 |
> exp/nnet3/tdnn_sp/decode_test/score_10_0.0/ctm.filt.filt.sys %WER 12.5 |
> 1155 27512 | 89.6 7.7 2.6 2.1 12.5 81.5 | -0.133 |
> exp/nnet3/tdnn_sp/decode_test_rescore/score_10_0.0/ctm.filt.filt.sys Training
> for 6 epochs.
>
> #%WER 15.2 | 507 17792 | 87.2 8.9 3.9 2.5 15.2 89.3 | -0.113 |
> exp/nnet3/tdnn_6epochs/decode_dev/score_10_1.0/ctm.filt.filt.sys
>
> #%WER 13.8 | 507 17792 | 89.0 8.1 2.9 2.9 13.8 87.0 | -0.144 |
> exp/nnet3/tdnn_6epochs/decode_dev_rescore/score_10_0.0/ctm.filt.filt.sys
>
> #%WER 14.0 | 1155 27512 | 88.4 8.8 2.8 2.4 14.0 84.2 | -0.161 |
> exp/nnet3/tdnn_6epochs/decode_test/score_10_0.0/ctm.filt.filt.sys
>
> #%WER 12.8 | 1155 27512 | 89.3 8.0 2.7 2.2 12.8 82.1 | -0.171 |
> exp/nnet3/tdnn_6epochs/decode_test_rescore/score_10_0.0/ctm.filt.filt.sys
>
> This might reflect the difference in transcription quality between
> {train,dev} and test sets; it could be overfitting to the poorer
> transcription. I will need to rerun the baseline without the messed up
> volume perturbation to be sure. (Waiting for the ivectors to be generated
> right now.)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/644#issuecomment-207902558>
>

galv commented:
Number of talks: 793 total.
```
$ wc -l data/{train,dev,test}/wav.scp
   774 data/train/wav.scp
     8 data/dev/wav.scp
    11 data/test/wav.scp
   793 total
```

Since you asked about speakers:
```
$ wc -l data/{train,dev,test}/spk2utt
    774 data/train/spk2utt
      8 data/dev/spk2utt
     11 data/test/spk2utt
    793 total
```

I know there are some repeats (eg., Bill Gates gave two talks in the dataset, so this speaker count is not totally correct). I doubt it matters, personally.

Need to do read the soxi man page to get durations... give me some time.

danpovey commented:
We apply all factors 0.9, 1, 1.1 to all conversations, so it doesn't matter
about the granularity.


On Sat, Apr 9, 2016 at 7:52 PM, Daniel Galvez <notifications@github.com>
wrote:

> Number of talks: 793 total.
>
> $ wc -l data/{train,dev,test}/wav.scp
>    774 data/train/wav.scp
>      8 data/dev/wav.scp
>     11 data/test/wav.scp
>    793 total
>
> Since you asked about speakers:
>
> $ wc -l data/{train,dev,test}/spk2utt
>     774 data/train/spk2utt
>       8 data/dev/spk2utt
>      11 data/test/spk2utt
>     793 total
>
> I know there are some repeats (eg., Bill Gates gave two talks in the
> dataset, so this speaker count is not totally correct). I doubt it matters,
> personally.
>
> Need to do read the soxi man page to get durations... give me some time.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/644#issuecomment-207903741>
>

vijayaditya commented:
I guess his concern is about volume perturbation which is randomly sampled
per recording
On Apr 9, 2016 23:23, "Daniel Povey" <notifications@github.com> wrote:

> We apply all factors 0.9, 1, 1.1 to all conversations, so it doesn't matter
> about the granularity.
>
>
> On Sat, Apr 9, 2016 at 7:52 PM, Daniel Galvez <notifications@github.com>
> wrote:
>
> > Number of talks: 793 total.
> >
> > $ wc -l data/{train,dev,test}/wav.scp
> > 774 data/train/wav.scp
> > 8 data/dev/wav.scp
> > 11 data/test/wav.scp
> > 793 total
> >
> > Since you asked about speakers:
> >
> > $ wc -l data/{train,dev,test}/spk2utt
> > 774 data/train/spk2utt
> > 8 data/dev/spk2utt
> > 11 data/test/spk2utt
> > 793 total
> >
> > I know there are some repeats (eg., Bill Gates gave two talks in the
> > dataset, so this speaker count is not totally correct). I doubt it
> matters,
> > personally.
> >
> > Need to do read the soxi man page to get durations... give me some time.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly or view it on GitHub
> > <https://github.com/kaldi-asr/kaldi/pull/644#issuecomment-207903741>
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/644#issuecomment-207906298>
>

galv commented:
Yes, the concern was volume perturbation. Still taking time to get the durations...

soxi does not accept piped arguments as input, unfortunately.

vijayaditya commented:
BTW you can use Sox -n stat to get durations.
On Apr 9, 2016 23:35, "Daniel Galvez" <notifications@github.com> wrote:

> Yes, the concern was volume perturbation. Still taking time to get the
> durations...
>
> soxi does not accept piped arguments as input, unfortunately.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/644#issuecomment-207910850>
>

galv commented:
For posterity's sake:

```
cat data/{train,dev,test}/wav.scp | \
  awk '{for (i=2; i<NF; i++) printf $i " "; print gensub(/.*\/([^.]+)\.sph/, "waves/\\1\.wav", "g", $(NF-1)) }' > sph2wav.sh
```

Creates a file that creates wav files. Then `soxi -T -d waves/*` can be run.

Total hours of all data is 210 hours.

From eye-balling, talks go from 6 to 25 minutes.

jtrmal commented:
IIRC the ted talks are usually 15 minutes.
On Apr 9, 2016 11:35 PM, "Daniel Galvez" <notifications@github.com> wrote:

> Yes, the concern was volume perturbation. Still taking time to get the
> durations...
>
> soxi does not accept piped arguments as input, unfortunately.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/644#issuecomment-207910850>
>

vijayaditya commented:
Is this ready for review ?

galv commented:
Nope; CUDNN has been occupying a lot of my time, for better or worse, and
my alptop has had terrible problems install NVIDIA drivers the past few
days...

Thanks for the reminder.

I may choose to retain the _sp suffix link Xingyu Na; removing it caused
some confusing (but fixed) errors, especially since we need _sp for the
data directories anyway.

On Tue, Apr 19, 2016 at 2:15 PM, Vijayaditya Peddinti <
notifications@github.com> wrote:

> Is this ready for review ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/644#issuecomment-212051780>
>



-- 
Daniel Galvez

danpovey commented:
I am OK with retaining the _sp suffix if removing it is becoming a hassle.
Dan


On Tue, Apr 19, 2016 at 1:12 PM, Daniel Galvez <notifications@github.com>
wrote:

> Nope; CUDNN has been occupying a lot of my time, for better or worse, and
> my alptop has had terrible problems install NVIDIA drivers the past few
> days...
>
> Thanks for the reminder.
>
> I may choose to retain the _sp suffix link Xingyu Na; removing it caused
> some confusing (but fixed) errors, especially since we need _sp for the
> data directories anyway.
>
> On Tue, Apr 19, 2016 at 2:15 PM, Vijayaditya Peddinti <
> notifications@github.com> wrote:
>
> > Is this ready for review ?
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly or view it on GitHub
> > <https://github.com/kaldi-asr/kaldi/pull/644#issuecomment-212051780>
> >
>
>
>
> --
> Daniel Galvez
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/644#issuecomment-212107162>
>

galv commented:
Removal of volume perturbation does not help. Decode results:

%WER 14.5 | 507 17792 | 88.8 7.8 3.4 3.3 14.5 94.5 | 0.064 | exp/chain/tdnn_no_vol/decode_dev/score_9_0.5/ctm.filt.filt.sys
%WER 13.6 | 507 17792 | 89.5 6.9 3.6 3.1 13.6 92.7 | 0.005 | exp/chain/tdnn_no_vol/decode_dev_rescore/score_10_0.0/ctm.filt.filt.sys

I am going to run an experiment again without my bug in data creation where I perturbed the test and dev sets. After this, I realistically should investigate differences in the tdnn models. But first I would like to get this PR in.

vimalmanohar commented:
It looks like padding + cleanup + combine_short_segments gives around 1% improvement over the baseline system from this pull request.
```
%WER 12.3 | 1155 27512 | 90.2 6.6 3.2 2.5 12.3 82.3 | 0.042 | exp/chain/tdnn_sp_cleaned_a_min1.55/decode_test_rescore/score_10_0.5/ctm.filt.filt.sys
%WER 12.4 | 507 17792 | 90.0 6.7 3.3 2.3 12.4 88.8 | -0.011 | exp/chain/tdnn_sp_cleaned_a_min1.55/decode_dev_rescore/score_10_0.5/ctm.filt.filt.sys
%WER 13.1 | 507 17792 | 89.4 7.5 3.1 2.5 13.1 89.0 | 0.092 | exp/chain/tdnn_sp_cleaned_a_min1.55/decode_dev/score_10_0.5/ctm.filt.filt.sys
%WER 13.3 | 1155 27512 | 89.3 7.6 3.0 2.6 13.3 84.7 | 0.120 | exp/chain/tdnn_sp_cleaned_a_min1.55/decode_test/score_10_0.5/ctm.filt.filt.sys
%WER 13.4 | 507 17792 | 89.8 6.9 3.3 3.2 13.4 93.3 | 0.036 | exp/chain/tdnn_sp/decode_dev_rescore/score_10_0.0/ctm.filt.filt.sys
%WER 13.5 | 1155 27512 | 90.0 6.6 3.3 3.5 13.5 90.6 | 0.057 | exp/chain/tdnn_sp/decode_test_rescore/score_10_0.0/ctm.filt.filt.sys
%WER 14.2 | 1155 27512 | 88.8 7.5 3.7 3.0 14.2 90.2 | 0.095 | exp/chain/tdnn_sp/decode_test/score_9_1.0/ctm.filt.filt.sys
%WER 14.3 | 507 17792 | 88.6 7.5 3.9 3.0 14.3 93.9 | 0.067 | exp/chain/tdnn_sp/decode_dev/score_9_1.0/ctm.filt.filt.sys
```

galv commented:
Which commit hash did you use? I found a small improvement from no longer using the speed-perturbed alignments in steps/nnet3/chain/build_tree.sh, but it's not up yet.

I had't posted yet because I was also fixing up scripts with the removal of the _sp suffix and I am waiting on discriminative training.

vimalmanohar commented:
This is the commit I used `036396efc35ad2cdb5b2039fdfce6b8e9f3d5786`. It was the latest in your tedlium-chain-pr branch.

galv commented:
It's interesting because when I tried combine_short_segments.py (by itself, since I believe cleanup takes a long time), I recall not getting a big improvement. In fact, it was a small decrease, I think.

I am curious how many bad utterances data clean up removed. You can just give me your experimental directory if you don't want to look into it yourself.

vimalmanohar commented:
I probably should have made it more clear. My cleanup here involves further steps beyond find_bad_utts.sh, like removing sub-segments of the utterance and not the entire utterance. I also modified the LM used in find_bad_utts.sh etc. The script is in local/chain/run_tdnn_cleanup.sh.
See `/home/vmanoha1/workspace_conf/egs/tedlium/s5/exp/tri3_ali_sp_train_padded_bad_utts/lattice_oracle/all_info.sorted.txt` for utterance WERs.

danpovey commented:
Great news!
Let us know what the timeline might be for getting this checked in-- I'm
not sure which scripts you used for this result?
Dan


On Thu, Apr 28, 2016 at 12:45 PM, Vimal Manohar <notifications@github.com>
wrote:

> It looks like padding + cleanup + combine_short_segments gives around 1%
> improvement over the baseline system from this pull request.
>
> %WER 12.3 | 1155 27512 | 90.2 6.6 3.2 2.5 12.3 82.3 | 0.042 | exp/chain/tdnn_sp_cleaned_a_min1.55/decode_test_rescore/score_10_0.5/ctm.filt.filt.sys
> %WER 12.4 | 507 17792 | 90.0 6.7 3.3 2.3 12.4 88.8 | -0.011 | exp/chain/tdnn_sp_cleaned_a_min1.55/decode_dev_rescore/score_10_0.5/ctm.filt.filt.sys
> %WER 13.1 | 507 17792 | 89.4 7.5 3.1 2.5 13.1 89.0 | 0.092 | exp/chain/tdnn_sp_cleaned_a_min1.55/decode_dev/score_10_0.5/ctm.filt.filt.sys
> %WER 13.3 | 1155 27512 | 89.3 7.6 3.0 2.6 13.3 84.7 | 0.120 | exp/chain/tdnn_sp_cleaned_a_min1.55/decode_test/score_10_0.5/ctm.filt.filt.sys
> %WER 13.4 | 507 17792 | 89.8 6.9 3.3 3.2 13.4 93.3 | 0.036 | exp/chain/tdnn_sp/decode_dev_rescore/score_10_0.0/ctm.filt.filt.sys
> %WER 13.5 | 1155 27512 | 90.0 6.6 3.3 3.5 13.5 90.6 | 0.057 | exp/chain/tdnn_sp/decode_test_rescore/score_10_0.0/ctm.filt.filt.sys
> %WER 14.2 | 1155 27512 | 88.8 7.5 3.7 3.0 14.2 90.2 | 0.095 | exp/chain/tdnn_sp/decode_test/score_9_1.0/ctm.filt.filt.sys
> %WER 14.3 | 507 17792 | 88.6 7.5 3.9 3.0 14.3 93.9 | 0.067 | exp/chain/tdnn_sp/decode_dev/score_9_1.0/ctm.filt.filt.sys
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/644#issuecomment-215490575>
>

galv commented:
I spent the last few days making sure there were no bugs with removing the _sp suffix from directories, but had not pushed that change yet. In addition, I incorporated a change from Vijay to train the chain tree using non-speed-perturbed alignments, which was helpful, and removed some dead code caught by static analysis by shellcheck.

Vimal's work is based on a commit that still uses the _sp suffix. For simplicity, I'll commit only my changes that don't change the _sp suffix; it'll be a headache to get it working with Vimal's new scripts othewise.

I'm not sure of the dependencies of Vimal's scripts, e.g., if everything needed in steps/ and utils/ is in master already.

danpovey commented:
Daniel, I think the parts of Vimal's work  relating to data cleanup may be
slightly longer-term, bcause those scripts will require a lot of review,
and he may not have 100% finalized the approach.  Just FYI-- don't plan on
this being merged in the next week or two.
Dan


On Thu, Apr 28, 2016 at 3:17 PM, Daniel Galvez <notifications@github.com>
wrote:

> I spent the last few days making sure there were no bugs with removing the
> _sp suffix from directories, but had not pushed that change yet. In
> addition, I incorporated a change from Vijay to train the chain tree using
> non-speed-perturbed alignments, which was helpful, and removed some dead
> code caught by static analysis by shellcheck.
>
> Vimal's work is based on a commit that still uses the _sp suffix. For
> simplicity, I'll commit only my changes that don't change the _sp suffix;
> it'll be a headache to get it working with Vimal's new scripts othewise.
>
> I'm not sure of the dependencies of Vimal's scripts, e.g., if everything
> needed in steps/ and utils/ is in master already.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/644#issuecomment-215533607>
>

vimalmanohar commented:
My code will be checked-in using a separate PR for data cleanup with recipes for different corpora. This will still take a few weeks. 
So Daniel can commit his changes without regard to the dependencies of my scripts. 
I just thought it appropriate reporting results here since chain system wasn't performing very well on Tedlium. Transcription errors might be the reason.

galv commented:
This is ready for rereview. I made a new commit  	82361ee at the end which shows my most recent changes for ease of review; **it is fine to review only this one**. We can squash later if desired.

I had to rebase my branch since it was getting out of date; I was worried some upstream change may cause a problem. That's why you see all commits listed at the bottom of this page.

vimalmanohar commented on egs/tedlium/s5/local/chain/run_tdnn.sh:
```diff
@@ -27,8 +27,7 @@ affix=
 get_egs_stage=-10
-speed_perturb=true
-dir=exp/chain/tdnn  # Note: _sp will get added to this if $speed_perturb == true.
+dir=exp/chain/tdnn$affix
```
$affix here will be empty and gets a value only after the parse_options.sh line even if the user supplied --affix option.
So the affix should be added to $dir after the parse_options.sh line.

vimalmanohar commented on egs/tedlium/s5/local/chain/run_tdnn.sh:
```diff
@@ -67,35 +65,25 @@ EOF
+  --speed-perturb true || exit 1;
 
 if [ $stage -le 9 ]; then
   # Get the alignments as lattices (gives the CTC training more freedom).
```
"chain" training

vimalmanohar commented on egs/tedlium/s5/local/chain/run_tdnn.sh:
```diff
@@ -67,35 +65,25 @@ EOF
 local/nnet3/run_ivector_common.sh --stage $stage \
-  --speed-perturb $speed_perturb \
-  --generate-alignments $speed_perturb || exit 1;
+  --speed-perturb true || exit 1;
```
I think you can pass --generate-alignments false because you don't need the alignments.

galv replied:
I was generating alignments because the baseline cross entropy nnet3 model uses the generated alignments anyway. I'll make a comment that it's optional.

vimalmanohar commented on egs/tedlium/s5/local/nnet3/run_ivector_common.sh:
```diff
@@ -7,62 +7,56 @@
+
       # this next section does volume perturbation on the data.
     cat $data_dir/wav.scp | python -c "
 import sys, os, subprocess, re, random
```
You have to pass a seed to the random number generator to ensure reproducibility. Use random.seed(0)

galv replied:
Did this, but I did not rerun the experiments.

vimalmanohar commented on egs/tedlium/s5/local/nnet3/run_ivector_common.sh:
```diff
@@ -76,38 +70,55 @@ for line in sys.stdin.readlines():
 if [ $stage -le 6 ]; then
   steps/online/nnet2/train_diag_ubm.sh --cmd "$train_cmd" --nj 30 --num-frames 700000 \
-    data/${train_set}_hires 512 exp/nnet3/tri4 exp/nnet3/diag_ubm
+    data/${train_set_sp}_hires 512 exp/nnet3/tri3b exp/nnet3/diag_ubm
```
${train_set}_sp_hires

galv commented:
Incorporated all of Vimal's comments by amending the previous commit. Thanks!

vijayaditya commented on egs/tedlium/s5/local/chain/README:
```diff
@@ -0,0 +1,17 @@
+These are instruction to reproduce the TEDLIUM models described in
```
instruction -> the instructions

vijayaditya commented on egs/tedlium/s5/local/chain/README:
```diff
@@ -0,0 +1,17 @@
+
+local/chain/run_tdnn.sh
+
+to see results for a generic chain model (these were not the results
```
we will be updating the paper with the latest results so the comment in brackets is unnecessary.

vijayaditya commented on egs/tedlium/s5/local/chain/run_tdnn.sh:
```diff
@@ -0,0 +1,205 @@
+# To reproduce the setup used in the paper, set the following variables:
+# affix=_more_ce
+# relu_dim=525
+# xent_regularize=0.2
```
It might be good to report the number of parameters in the chain portion of the neural network and chain + xent portion of the neural network.

galv replied:
What do you mean by "portion"? Do you mean you would like to see the number of parameters of each of the two models whose results I provide?

vijayaditya commented on egs/tedlium/s5/local/chain/run_tdnn.sh:
```diff
@@ -0,0 +1,205 @@
+# allows us to run local/nnet3 models without having to run this script first.
+# If you don't want to run those models, you can do "--generate-alignments false"
+local/nnet3/run_ivector_common.sh --stage $stage \
+  --generate-alignments true \
```
we usually run just the best model so use `--generate-alignments false` as the default.

vijayaditya commented on egs/tedlium/s5/local/chain/run_tdnn.sh:
```diff
@@ -0,0 +1,205 @@
+# nnet3 setup, and you can skip them by setting "--stage 9" if you have already
+# run those things.
+
+train_set=train
```
the variable `train_set` is redundant.

vijayaditya commented:
Please squash the commits. If you are not able to do so, because you did not rebase your local branch, I would recommend migrating these changes to a new branch (should be easy if you create a patch and apply it on the new branch). 

When you create the new PR please refer to this PR to preserve the details of this discussion.

vijayaditya commented on egs/tedlium/s5/local/chain/run_tdnn.sh:
```diff
@@ -0,0 +1,205 @@
+  for decode_set in dev test; do
+    (
+    steps/nnet3/decode.sh --acwt 1.0 --post-decode-acwt 10.0 \
+      --nj 8 --cmd "$decode_cmd" $iter_opts \
```
It would be better to determine the number of jobs based on the number of speakers in the dataset.

galv replied:
There are 8 speakers in dev set and 11 in test set. Do you mean that I should get the number of speakers dynamically by calling something like `$(wc -l data/dev/spk2utt)`?

galv replied:
@vijayaditya Answer this and I'll get started on the new PR.

By the way, sorry for nonreponsiveness; now is the time of final projects galore :(

vijayaditya replied:
Yes.

BTW if you are busy with final projects this can wait. People who want to work on Ted-LIUM nnet3 could always clone your PR branch locally.

galv replied:
I actually already incorporated most of your suggestions, but forgot about this when you didn't respond to this. I'll take a look at squashing and PRing again hopefully within the hour. I doubt there'll be any problems. I'm a magit user!

vijayaditya commented:
@danpovey looks good to me. Can be merged once the commits have been squashed and the results have been updated.

danpovey commented:
OK- merge yourself when it looks OK to you; the same with similar commits.


On Sat, May 7, 2016 at 2:47 PM, Vijayaditya Peddinti <
notifications@github.com> wrote:

> @danpovey <https://github.com/danpovey> looks good to me. Can be merged
> once the commits have been squashed and the results have been updated.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/644#issuecomment-217661286>
>

galv commented:
When I open a PR, I won't have squashed 688802b4c92b0b956cd760e83eeaa96e8e1c9871. It deserves its own commit. Just a heads up.

vijayaditya commented:
@galv Whats the status of the PR with the squashed commits ? Could you get this done today ? There are some other PRs which depend on this merge.

galv commented:
Final commit 362f69f can be seen here, before I squashed all the commits.
